_posts/2015-06-09-Fostering-Creativity-with-Semantically-Rich-APIs.md

```diff
@@ -0,0 +1,115 @@
+---
+title: Semantically Rich APIs
+excerpt: Fostering Creativity with Semantically Rich APIs
+layout: post
+date: 2015-06-09 00:00:00 +0200
+author: stephen
+published: true
+---
+
+In our recent blog post, Kyle Fuller [wrote about](http://blog.apiary.io/2015/06/04/Hyperdrive/) the benefits of decoupling semantics from implementation details for building better API clients. I'd like to explore how this decoupled approach also benefits the API design process—it's quite nice to decouple your mind from implementation details as well.
+
+## The Difficulty with Common Approaches
+
+Many modern approaches in designing and building APIs couple not only clients and servers to implementation details, but also couple thought, creativity, and shared understanding to that implementation. Coupling servers and clients is a problem that is easily seen, and in theory no one wants a coupled approach. Kyle Fuller wrote in his post:
+
+> Imagine entering an API without knowing about any implementation details, but instead just understanding the semantics of the API. We can get the implementation details of the API at run-time to allow us to change them at any point.
+
+In waiting until runtime to get these specific details, Hyperdrive allows for building clients that are highly evolvable and don't break when things are changed on the server.
+
+Coupling on the other hand hinders this API evolvability—among other things—but it also hinders API design and creativity for designers and consumers. I think this is something we miss. We're all happy to have won the battle over technologies like SOAP and WSDL, but we fail to see our tools and methodologies are leading us right back to where we were.
+
+## Chewing the CRUD
+
+If you survey the modern API space, you'll see this coupled design pattern is the norm, albeit it's a limited one. The first design step in this pattern is usually to carefully craft a URL that conveys to the users the meaning of the content. The next step is to pick the HTTP method the users will use in interacting with that URL. The thought, design, and creativity is directly coupled with the HTTP implementation in this process, and pushes the question "What does my API do?" to later steps in the design process.
```

Comment:
I am afraid many people maps "What does my API do?" to "URI design" so to them it will be the false statement (Hey I am thinking what my API is doing that is why I pick URI verbs / nouns so carefully). 

I think the real problems goes back to the coupling or more precisely the **lack of understanding what coupling is** and what are the consequences of it...